### PR TITLE
separate namespace for each organization

### PR DESCRIPTION
**Pull request by basilky**
Currently all the organizations are in a single namespace. Changes are done in multiple files to create three namespaces and to separate organizations across namespaces.
- ORG0_NS, ORG1_NS, ORG3_NS are defined in network script
- $NS in the scripts are replaced with corresponding namespace

All till chaincode invoke/query is working. I'm yet to test the api server. kind was used for n/w setup, will test on Rancher also.

fixes https://github.com/hyperledger/fabric-samples/issues/644.

Signed-off-by: Basil K Y <techiebasil@gmail.com>

### CODE REVIEW

**Comment by jkneubuh** [comment]
Wow.  This is great.  Thank you, @basilky ! 

A couple of questions / comments: 

- This is great. Thank you for the contribution. 

- Are the ci/ test script failures related to the PR?  Sometimes the tests are "flaky" at AZP.  ( running `fabric-samples/ci/scripts/run-k8s-test-network-basic.sh` locally is a good way to suss out errors from CI ) 

- The Kube Test Network is showing some ... "age."   It is good for a reference for deploying to k8s, or for illustrating different techniques and testing integrations (e.g. cert-manager.io or linkerd.io for a mesh).    Have you seen the new fabric-operator and associated [sample-network](https://github.com/hyperledger-labs/fabric-operator/tree/main/sample-network) ?    It is similar to the kube test network, but ... better in some ways.


Note I am not suggesting this PR shouldn't be merged.  Only that it's worth taking a look at fabric-operator, which is "production grade," following similar patterns as exposed by the kube test network.

**Review by jkneubuh — CHANGES_REQUESTED** [review]
I think this is a really great PR.  Thanks again for moving it forward.   A couple of asks / requests :

1.  Patch up the AZP test runs.  (Not clear if this is an AZP test flake, or possibly related to the update to span multiple namespaces.)   The k8s tests under "/ci" should be enough to drive the overall flow. 

2.  Why was the change of working directory for the channel config necessary?  (Moving ../../build/channel-msp --> ./channel-msp)    I recall arguing with this area of the code / setup, and wasn't very happy embedding the working directory for the channel config block.

3.  Can we leave the existing behavior of deploying to a single namespace as the "default" behavior (e.g. -> `test-network`) and then use this PR + a guide to instruct on deployment of the different orgs to multiple namespaces?


This looks great.

**Comment by basilky** [comment]
> I think this is a really great PR. Thanks again for moving it forward. A couple of asks / requests :
> 
>     1. Patch up the AZP test runs.  (Not clear if this is an AZP test flake, or possibly related to the update to span multiple namespaces.)   The k8s tests under "/ci" should be enough to drive the overall flow.
> 
>     2. Why was the change of working directory for the channel config necessary?  (Moving ../../build/channel-msp --> ./channel-msp)    I recall arguing with this area of the code / setup, and wasn't very happy embedding the working directory for the channel config block.
> 
>     3. Can we leave the existing behavior of deploying to a single namespace as the "default" behavior (e.g. -> `test-network`) and then use this PR + a guide to instruct on deployment of the different orgs to multiple namespaces?
> 
> 
> This looks great.

@jkneubuh Thanks for the review and suggestions.

1. I ran the CI script multiple times locally and the results are inconsistent. Sometimes it fails due to kubectl timeout, sometimes due to ssl handshake error, sometimes docker deamon failing to kill kind cluster etc. Will try to make it work.

2. I had added the namespace variable into the ordering and peers endpoints in configtx.yaml(and renamed configtx.yaml to configtx-template.yaml). `envsubst` is used to generate actual configtx.yaml from this template. The generated configtx.yaml is saved to the build directory. So I changed "../../build/" to "./" as configtx.yaml is already inside build directory. 

3. I will change all three namespace variables to $NS so that all organizations will be in a single namespace by default.

**Comment by jkneubuh** [comment]
Thanks @basilky .   Holler if you are stuck on the test suites.  I agree they are ... "brittle."  (More reason / incentive to use the fabric-operator, IMO! ) 

Thanks for moving the configtx to a template / etc.  I understand the update now.  I think there may still be some lingering overlap between what runs "locally" vs. running in a clean sandbox, as is the case in the CI flow when everything starts up from scratch.

Also, I think there may be a couple of lingering issues where multiple "network up" calls will re-use the certificates and enrollments generated from a previous run of the network.   For instance, if you delete a kind cluster with the KIND cli, this will not scrub the old certs, and the next network to spin up will be doomed from the start.

I _think_ there may be a case in here where there is still some overlap between the configtx.yaml downloaded when the peer binaries are installed by the network script, and the template substitutions above.  Holler if you want another pair of eyes to look at the test scripts - I agree these are very ... fortunate to be passing at all.   Lots of room for improvement in here.

Thanks again for the PR.

File: test-network-k8s/kube/fabric-rest-sample.yaml
Comment (jkneubuh) [inline test-network-k8s/kube/fabric-rest-sample.yaml:120]: I think this is a "bug" but is OK in this context.  It's not worth holding up this PR for a fix.

When connecting with the new Gateway Client, the application can connect to ANY of the peers in the org, and the gateway peer will act as a proxy or tunnel to the actual peer instance that is invoking the transactions on the ledger.   In this case, we set up a Kube `Service`  - `org1-peer-gateway-svc` to distribute client connections across the peers in the org.    The URLs here will work fine, but can probably be improved through access via the Gateway / virtual URLs.  Good examples of this feature in action: 

- [Fabric Gateway](https://github.com/hyperledgendary/full-stack-asset-transfer-guide/blob/main/docs/ApplicationDev/01-FabricGateway.md)

- [Gateway Load Balancing](https://github.com/hyperledgendary/full-stack-asset-transfer-guide/blob/main/docs/CloudReady/40-bananas.md#gateway-load-balancing)

**Comment by jkneubuh** [comment]
This PR looks great.  @basilky -- I'm more than happy to land this when you are satisfied and switch the PR out of a DRAFT state.  

👍

File: test-network-k8s/kube/fabric-rest-sample.yaml
Comment (basilky) [reply test-network-k8s/kube/fabric-rest-sample.yaml]: @jkneubuh I have reverted this change and the apis are working

**Comment by basilky** [comment]
@jkneubuh I think this PR is ready to merge. Let me know if any changes are needed. btw, the fabric-operator is new to me. I will definitely have a look at it.

**Comment by jkneubuh** [comment]
Thanks @basilky ! 

Start with an introduction to the operator at https://github.com/hyperledgendary/full-stack-asset-transfer-guide 

The guide is still a little light on words / figures / explanatory text.  Paired with the operator sample-network I think there is enough here to convince you to stop improving the Kube Test Network.  : ) 

The "Cloud Native Fabric" section has the most graceful introduction, IMO.   It's "all here" - minus the cert-manager.io integration -- fabric-operator is a much, much, much better path for deploying production-ready Fabric networks.


Thank you for the great PR.  Cheers.

**Review by jkneubuh — APPROVED** [review]

**Review by jkneubuh — APPROVED** [review]

**Pull request merged**
